### PR TITLE
provide more information on kernel compilation error

### LINKED ISSUE #2119: More informative errors when parameter size is too big
**Is your feature request related to a problem? Please describe.**

If my parameter size is too big, I get a rather complicated error:
```
ERROR: Failed to compile PTX code (ptxas exited with code 255)
Invocation arguments: --generate-line-info --verbose --gpu-name sm_80 --output-file /tmp/jl_o5IvWTUhWn.cubin /tmp/jl_cRUv2pSP6J.ptx
ptxas /tmp/jl_cRUv2pSP6J.ptx, line 147; error   : Entry function '_Z16broadcast_kernel15CuKernelContext13CuDeviceArrayI7Float64Li1ELi1EE11BroadcastedI12CuArrayStyleILi1EE5TupleI5OneToI5Int64EE1fS4_I8ExtrudedIS0_IS1_Li1ELi1EES4_I4BoolES4_IS6_EE10CuRefValueIS4_IS1_S1_S1_S1_S1_S
#
# lots of stuff
#
uses too much parameter space (0x138f0 bytes, 0x7ffc max).
ptxas fatal   : Ptx assembly aborted due to errors
If you think this is a bug, please file an issue and attach /tmp/jl_cRUv2pSP6J.ptx
```

**Describe the solution you'd like**

It would be nice if we could get a more informative error, perhaps giving the size recursively of each parameter passed to the kernel, similar to the non-`isbits` error.

**Describe alternatives you've considered**

**Additional context**

I'm not sure if there is some way to query the maximum parameter size? According to https://developer.nvidia.com/blog/cuda-12-1-supports-large-kernel-parameters/, it depends on both the CUDA runtime and architecture.

### PR DESCRIPTION
**Pull request by simonbyrne**
Fixes #2119.

### CODE REVIEW

**Comment by simonbyrne** [comment]
Example:
```
┌ Error: Error compiling CUDA kernel rte_sw_2stream_solve_CUDA! with args: 
│   flux :: RRTMGP.Fluxes.FluxSW{Float64, CUDA.CuDeviceMatrix{Float64, 1}}   (160 bytes)
│   flux_sw :: RRTMGP.Fluxes.FluxSW{Float64, CUDA.CuDeviceMatrix{Float64, 1}}   (160 bytes)
│   op :: RRTMGP.Optics.TwoStream{Float64, CUDA.CuDeviceMatrix{Float64, 1}}   (120 bytes)
│   bcs_sw :: RRTMGP.BCs.SwBCs{Float64, CUDA.CuDeviceVector{Float64, 1}, Nothing, CUDA.CuDeviceMatrix{Float64, 1}}   (144 bytes)
│   src_sw :: RRTMGP.Sources.SourceSW2Str{Float64, CUDA.CuDeviceVector{Float64, 1}, CUDA.CuDeviceMatrix{Float64, 1}}   (392 bytes)
│   nlay :: Int64   (8 bytes)
│   ncol :: Int64   (8 bytes)
│   major_gpt2bnd :: CUDA.CuDeviceVector{UInt8, 1}   (32 bytes)
│   solar_src_scaled :: CUDA.CuDeviceVector{Float64, 1}   (32 bytes)
│   as :: RRTMGP.AtmosphericStates.AtmosphericState{Float64, CUDA.CuDeviceVector{Float64, 1}, CUDA.CuDeviceVector{Float64, 1}, CUDA.CuDeviceMatrix{Float64, 1}, CUDA.CuDeviceMatrix{Float64, 1}, CUDA.CuDeviceMatrix{Bool, 1}, CUDA.CuDeviceMatrix{Float64, 1}, RRTMGP.AtmosphericStates.MaxRandomOverlap, RRTMGP.Vmrs.VmrGM{Float64, CUDA.CuDeviceVector{Float64, 1}, CUDA.CuDeviceMatrix{Float64, 1}}}   (800 bytes)
│   lookup_sw :: RRTMGP.LookUpTables.LookUpSW{Float64, CUDA.CuDeviceVector{UInt8, 1}, CUDA.CuDeviceVector{Int64, 1}, CUDA.CuDeviceMatrix{Int64, 1}, CUDA.CuDeviceArray{Int64, 3, 1}, CUDA.CuDeviceVector{Float64, 1}, CUDA.CuDeviceMatrix{Float64, 1}, CUDA.CuDeviceArray{Float64, 3, 1}, CUDA.CuDeviceArray{Float64, 4, 1}}   (1320 bytes)
│   lookup_sw_cld :: RRTMGP.LookUpTables.LookUpCld{Bool, Float64, CUDA.CuDeviceVector{Float64, 1}, CUDA.CuDeviceMatrix{Float64, 1}, CUDA.CuDeviceArray{Float64, 3, 1}, CUDA.CuDeviceArray{Float64, 4, 1}}   (936 bytes)
└ @ RRTMGP.RTESolver /central/home/spjbyrne/src/RRTMGP.jl/src/rte/RTESolver.jl:445
ERROR: LoadError: CUDA error: device kernel image is invalid (code 200, ERROR_INVALID_IMAGE)

Stacktrace:
  [1] CUDA.CuModule(data::Vector{UInt8}, options::Dict{CUDA.CUjit_option_enum, Any})
    @ CUDA /central/home/spjbyrne/src/CUDA.jl/lib/cudadrv/module.jl:57
  [2] CuModule
    @ /central/home/spjbyrne/src/CUDA.jl/lib/cudadrv/module.jl:23 [inlined]
  [3] link(job::GPUCompiler.CompilerJob, compiled::NamedTuple{(:image, :entry, :external_gvars), Tuple{Vector{UInt8}, String, Vector{String}}})
    @ CUDA /central/home/spjbyrne/src/CUDA.jl/src/compiler/compilation.jl:365
  [4] actual_compilation(cache::Dict{Any, CUDA.CuFunction}, src::Core.MethodInstance, world::UInt64, cfg::GPUCompiler.CompilerConfig{GPUCompiler.PTXCompilerTarget, CUDA.CUDACompilerParams}, compiler::typeof(CUDA.compile), linker::typeof(CUDA.link))
    @ GPUCompiler ~/.julia/packages/GPUCompiler/U36Ed/src/execution.jl:132
  [5] cached_compilation(cache::Dict{Any, CUDA.CuFunction}, src::Core.MethodInstance, cfg::GPUCompiler.CompilerConfig{GPUCompiler.PTXCompilerTarget, CUDA.CUDACompilerParams}, compiler::Function, linker::Function)
    @ GPUCompiler ~/.julia/packages/GPUCompiler/U36Ed/src/execution.jl:103
  [6] macro expansion
    @ /central/home/spjbyrne/src/CUDA.jl/src/compiler/execution.jl:392 [inlined]
  [7] macro expansion
    @ ./lock.jl:267 [inlined]
  [8] cufunction(f::typeof(RRTMGP.RTESolver.rte_sw_2stream_solve_CUDA!), tt::Type{Tuple{RRTMGP.Fluxes.FluxSW{Float64, CUDA.CuDeviceMatrix{Float64, 1}}, RRTMGP.Fluxes.FluxSW{Float64, CUDA.CuDeviceMatrix{Float64, 1}}, RRTMGP.Optics.TwoStream{Float64, CUDA.CuDeviceMatrix{Float64, 1}}, RRTMGP.BCs.SwBCs{Float64, CUDA.CuDeviceVector{Float64, 1}, Nothing, CUDA.CuDeviceMatrix{Float64, 1}}, RRTMGP.Sources.SourceSW2Str{Float64, CUDA.CuDeviceVector{Float64, 1}, CUDA.CuDeviceMatrix{Float64, 1}}, Int64, Int64, CUDA.CuDeviceVector{UInt8, 1}, CUDA.CuDeviceVector{Float64, 1}, RRTMGP.AtmosphericStates.AtmosphericState{Float64, CUDA.CuDeviceVector{Float64, 1}, CUDA.CuDeviceVector{Float64, 1}, CUDA.CuDeviceMatrix{Float64, 1}, CUDA.CuDeviceMatrix{Float64, 1}, CUDA.CuDeviceMatrix{Bool, 1}, CUDA.CuDeviceMatrix{Float64, 1}, RRTMGP.AtmosphericStates.MaxRandomOverlap, RRTMGP.Vmrs.VmrGM{Float64, CUDA.CuDeviceVector{Float64, 1}, CUDA.CuDeviceMatrix{Float64, 1}}}, RRTMGP.LookUpTables.LookUpSW{Float64, CUDA.CuDeviceVector{UInt8, 1}, CUDA.CuDeviceVector{Int64, 1}, CUDA.CuDeviceMatrix{Int64, 1}, CUDA.CuDeviceArray{Int64, 3, 1}, CUDA.CuDeviceVector{Float64, 1}, CUDA.CuDeviceMatrix{Float64, 1}, CUDA.CuDeviceArray{Float64, 3, 1}, CUDA.CuDeviceArray{Float64, 4, 1}}, RRTMGP.LookUpTables.LookUpCld{Bool, Float64, CUDA.CuDeviceVector{Float64, 1}, CUDA.CuDeviceMatrix{Float64, 1}, CUDA.CuDeviceArray{Float64, 3, 1}, CUDA.CuDeviceArray{Float64, 4, 1}}}}; kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ CUDA /central/home/spjbyrne/src/CUDA.jl/src/compiler/execution.jl:387
  [9] cufunction
    @ /central/home/spjbyrne/src/CUDA.jl/src/compiler/execution.jl:384 [inlined]
 [10] macro expansion
    @ /central/home/spjbyrne/src/CUDA.jl/src/compiler/execution.jl:105 [inlined]
 [11] rte_sw_2stream_solve!(context::ClimaComms.SingletonCommsContext{ClimaComms.CUDADevice}, flux::RRTMGP.Fluxes.FluxSW{Float64, CUDA.CuArray{Float64, 2, CUDA.Mem.DeviceBuffer}}, flux_sw::RRTMGP.Fluxes.FluxSW{Float64, CUDA.CuArray{Float64, 2, CUDA.Mem.DeviceBuffer}}, op::RRTMGP.Optics.TwoStream{Float64, CUDA.CuArray{Float64, 2, CUDA.Mem.DeviceBuffer}}, bcs_sw::RRTMGP.BCs.SwBCs{Float64, CUDA.CuArray{Float64, 1, CUDA.Mem.DeviceBuffer}, Nothing, CUDA.CuArray{Float64, 2, CUDA.Mem.DeviceBuffer}}, src_sw::RRTMGP.Sources.SourceSW2Str{Float64, CUDA.CuArray{Float64, 1, CUDA.Mem.DeviceBuffer}, CUDA.CuArray{Float64, 2, CUDA.Mem.DeviceBuffer}}, major_gpt2bnd::CUDA.CuArray{UInt8, 1, CUDA.Mem.DeviceBuffer}, solar_src_scaled::CUDA.CuArray{Float64, 1, CUDA.Mem.DeviceBuffer}, max_threads::Int64, as::RRTMGP.AtmosphericStates.AtmosphericState{Float64, CUDA.CuArray{Float64, 1, CUDA.Mem.DeviceBuffer}, CUDA.CuArray{Float64, 1, CUDA.Mem.DeviceBuffer}, CUDA.CuArray{Float64, 2, CUDA.Mem.DeviceBuffer}, CUDA.CuArray{Float64, 2, CUDA.Mem.DeviceBuffer}, CUDA.CuArray{Bool, 2, CUDA.Mem.DeviceBuffer}, CUDA.CuArray{Float64, 2, CUDA.Mem.DeviceBuffer}, RRTMGP.AtmosphericStates.MaxRandomOverlap, RRTMGP.Vmrs.VmrGM{Float64, CUDA.CuArray{Float64, 1, CUDA.Mem.DeviceBuffer}, CUDA.CuArray{Float64, 2, CUDA.Mem.DeviceBuffer}}}, lookup_sw::RRTMGP.LookUpTables.LookUpSW{Float64, CUDA.CuArray{UInt8, 1}, CUDA.CuArray{Int64, 1}, CUDA.CuArray{Int64, 2}, CUDA.CuArray{Int64, 3}, CUDA.CuArray{Float64, 1}, CUDA.CuArray{Float64, 2}, CUDA.CuArray{Float64, 3}, CUDA.CuArray{Float64, 4}}, lookup_sw_cld::RRTMGP.LookUpTables.LookUpCld{Bool, Float64, CUDA.CuArray{Float64, 1}, CUDA.CuArray{Float64, 2}, CUDA.CuArray{Float64, 3}, CUDA.CuArray{Float64, 4}})
    @ RRTMGP.RTESolver /central/home/spjbyrne/src/RRTMGP.jl/src/rte/RTESolver.jl:445
...
```

**Comment by simonbyrne** [comment]
thoughts?

**Review by vchuravy — COMMENTED** [review]
Looks great to me!

**Comment by nomadbl** [comment]
Here is a suggestion/question:
Here is the output in a case where I got the error:
```
┌ Error: Error compiling CUDA kernel kern_filter_residual_op! with args: 
│   dw :: TemplateMat{Float32, 70, CuDeviceVector{Float32, 1}}   (2240 bytes)
│   w :: TemplateMat{Float32, 70, CuDeviceVector{Float32, 1}}   (2240 bytes)
│   x :: CuDeviceMatrix{Float32, 1}   (40 bytes)
│   dy :: CuDeviceMatrix{Float32, 1}   (40 bytes)
└ @ OpeNet /app/OpeNet/src/resconv_im2col.jl:201
ERROR: Failed to compile PTX code (ptxas exited with code 255)
Invocation arguments: --generate-line-info --verbose --gpu-name sm_75 --output-file /tmp/jl_lMPGv0pAvO.cubin /tmp/jl_JCJcvgpKre.ptx
ptxas /tmp/jl_JCJcvgpKre.ptx, line 1478; error   : Entry function '_Z24kern_filter_residual_op_11TemplateMatI7Float32Li70E13CuDeviceArrayIS0_Li1ELi1EEES_IS0_Li70ES1_IS0_Li1ELi1EEES1_IS0_Li2ELi1EES1_IS0_Li2ELi1EE' uses too much parameter space (0x11e0 bytes, 0x1100 max).
ptxas fatal   : Ptx assembly aborted due to errors
```
Now, the reported overall use is 11e0=4576 bytes vs 4560 bytes summing up the individual variables.
Where do the extra 16 bytes come from?

Other than that, great and looking forward to a merge.

**Comment by simonbyrne** [comment]
yeah, @sriharshakandala had an example with something similar. I'm not sure?

**Comment by sriharshakandala** [comment]
> yeah, @sriharshakandala had an example with something similar. I'm not sure?

I had a case in which  the total metadata size, measured using the above criterion, is below 4 kb, but still crashed, for larger resolution version of a problem on P100s. 

The same problem, with smaller resolution, but same metadata size, ran correctly on P100.  It has been verified the GPU memory limit has not been exceeded. Both worked fine on a A100, which is expected.

I was wondering if the above criteria is a deterministic indicator of the kernel metadata size!

File: src/compiler/execution.jl [hunk 1]
```diff
@@ -101,7 +101,17 @@ macro cuda(ex...)
...
+                        $cufunction($kernel_f, $kernel_tt; $(compiler_kwargs...))
+                    catch err
+                        f_str = $(string(f))
+                        arg_strs = ["\n  $arg :: $type   ($(sizeof(type)) bytes)" for (arg, type) in zip($(map(string, args)), fieldtypes($kernel_tt))]
+                        @error(
+                            "Error compiling CUDA kernel $(f_str) with args: $(join(arg_strs))",
+                            _file=$(__source__.file === nothing ? "?" : String(__source__.file::Symbol)),
+                            _line=$(__source__.line))
+                        rethrow(err)
+                    end
```
Comment (maleadt) [inline src/compiler/execution.jl]: I'd rather we do this in the `link` function, where we can inspect the argument types from the method instance. This generates a whole lot of code now that's being inlined in every `@cuda` call site.

Also, it should probably only happen when relevant exceptions occur. The `CuError` has a metadata field for exactly that purpose: https://github.com/JuliaGPU/CUDA.jl/blob/f185f79ce57745bea1ca7ba913d3925772ad12a2/lib/cudadrv/module.jl#L57

**Comment by maleadt** [comment]
> Where do the extra 16 bytes come from?

Probably the KernelState argument, which is added by the compiler. The code should take that into account.

File: src/compiler/execution.jl (hunk 1 above)
Comment (maleadt) [reply src/compiler/execution.jl]: Or rather surrounding the `ptxas` invocation, I guess. Maybe we could also just hard-code the maximum parameter sizes that are allowed, which is even easier. It's 4096 bytes, except on Volta usin gCUDA 12.1 and a sufficiently high PTX target where it's 32764 bytes now.

**Comment by maleadt** [comment]
@simonbyrne Thoughts?

File: src/compiler/compilation.jl [hunk 2]
```diff
@@ -273,9 +273,23 @@ function compile(@nospecialize(job::CompilerJob))
...
+    argtypes = filter([KernelState, job.source.specTypes.parameters...]) do dt
+        !isghosttype(dt) && !Core.Compiler.isconstType(dt)
+    end
+    param_usage = sum(sizeof, argtypes)
+    param_limit = 4096
+    if cap >= v"7.0" && ptx >= v"8.1"
+        param_limit = 32764
+    end
+    if param_usage > param_limit
+        error("Kernel invocation uses $(Base.format_bytes(param_usage)) of parameters, exceeding the $(Base.format_bytes(param_limit)) limit.")
```
Comment (simonbyrne) [inline src/compiler/compilation.jl]: I was really hoping we could have a way to dump the size of each parameter, which we have found useful when trying to identify which argument is the cause of the failure (or perhaps even have a it log it always when using a debug mode?(

Comment (maleadt) [reply src/compiler/compilation.jl]: Sure, that's just a matter of tweaking the error message.

**Comment by maleadt** [comment]
Tweaked the output a little:

```
julia> @cuda kernel(ntuple(_->UInt64(1), 2^13))
ERROR: Kernel invocation uses too much parameter memory.
64.016 KiB exceeds the 31.996 KiB limit imposed by sm_89 / PTX v8.2.

Relevant parameters:
  [1] x::NTuple{8192, UInt64} uses 64.000 KiB

Stacktrace:
  [1] error(s::String)
    @ Base ./error.jl:35
```

**Comment by simonbyrne** [comment]
That looks great, thanks!

**Pull request merged**